diff --git a/usr/share/lipstick-jolla-home-qt5/layers/HomeLayer.qml b/usr/share/lipstick-jolla-home-qt5/layers/HomeLayer.qml
index 188bdf3..d13ac72 100644
--- a/usr/share/lipstick-jolla-home-qt5/layers/HomeLayer.qml
+++ b/usr/share/lipstick-jolla-home-qt5/layers/HomeLayer.qml
@@ -43,7 +43,7 @@ Pannable {
     readonly property bool _transposed: Lipstick.compositor.topmostWindowAngle % 180 != 0
 
     function partnerLayerForWindow(window) {
-        for (var layer = switcherLayer.rightItem; layer != eventsLayer; layer = layer.rightItem) {
+        for (var layer = switcherLayer.rightItem; layer && layer != eventsLayer; layer = layer.rightItem) {
             if (JollaSystemInfo.isWindowForLauncherItem(window, layer.launcherItem)) {
                 return layer
             }
@@ -52,7 +52,7 @@ Pannable {
     }
 
     function activatePartnerWindow(launcherItem, launch) {
-        for (var layer = switcherLayer.rightItem; layer != eventsLayer; layer = layer.rightItem) {
+        for (var layer = switcherLayer.rightItem; layer && layer != eventsLayer; layer = layer.rightItem) {
             if (!layer.launcherItem || layer.launcherItem.exec != launcherItem.exec) {
                 continue
             } else if (!layer.window.window && launch) {
@@ -81,7 +81,7 @@ Pannable {
 
     function _findNextLayer(launcherItem) {
         var index = partnerspaces.indexOf(launcherItem)
-        for (var layer = switcherLayer.rightItem; layer != eventsLayer; layer = layer.rightItem) {
+        for (var layer = switcherLayer.rightItem; layer && layer != eventsLayer; layer = layer.rightItem) {
             if (layer.launcherItem && index < partnerspaces.indexOf(layer.launcherItem)) {
                 return layer
             }
@@ -189,7 +189,7 @@ Pannable {
 
             objectName: "eventsLayer"
 
-            leftItem: switcherLayer
+            leftItem: null
             rightItem: switcherLayer
 
             width: homescreen.width
@@ -197,7 +197,7 @@ Pannable {
 
             visible: false
 
-            anchors { leftMargin: leftItem.minimizeMargin }
+            anchors { rightMargin: -rightItem.minimizeMargin }
         },
         SwitcherLayer {
             id: switcherLayer
@@ -217,7 +217,7 @@ Pannable {
             objectName: "switcherLayer"
 
             leftItem:  eventsLayer
-            rightItem: eventsLayer
+            rightItem: null
 
             width: homescreen.width
             height: homescreen.height
@@ -242,7 +242,7 @@ Pannable {
                 NumberAnimation { duration: 250; easing.type: Easing.OutQuad }
             }
 
-            anchors { rightMargin: -rightItem.minimizeMargin }
+            anchors { leftMargin: leftItem.minimizeMargin }
         }
     ]
 
@@ -266,7 +266,7 @@ Pannable {
         onItemRemoved: {
             var apkdPackage = item.readValue("X-apkd-packageName")
 
-            for (var layer = switcherLayer.rightItem; layer != eventsLayer; layer = layer.rightItem) {
+            for (var layer = switcherLayer.rightItem; layer && layer != eventsLayer; layer = layer.rightItem) {
                 if (layer.launcherItem == item) {
                     function cleanup() {
                         var left = layer.leftItem
@@ -306,7 +306,7 @@ Pannable {
             var i = count
             var firstMoved = eventsLayer
             var lastMoved = eventsLayer
-            for (var layer = switcherLayer.rightItem; layer != eventsLayer; layer = layer.rightItem) {
+            for (var layer = switcherLayer.rightItem; layer && layer != eventsLayer; layer = layer.rightItem) {
                 for (i = 0; i < count; ++i) {
                     if (layer.launcherItem == partnerspaces.get(index + i)) {
                         firstMoved = layer
diff --git a/usr/share/lipstick-jolla-home-qt5/main/EditableGridDelegate.qml b/usr/share/lipstick-jolla-home-qt5/main/EditableGridDelegate.qml
index c5e65e3..c0cd3e9 100644
--- a/usr/share/lipstick-jolla-home-qt5/main/EditableGridDelegate.qml
+++ b/usr/share/lipstick-jolla-home-qt5/main/EditableGridDelegate.qml
@@ -100,11 +100,14 @@ Item {
     MouseArea {
         id: delegateContentItem
 
+        property int oldZ
+
         function startReordering() {
             if (editMode && !dragged) {
                 reparent(manager.dragContainer)
                 manager.reorderItem = delegateContentItem
                 drag.target = delegateContentItem
+                oldZ = z
                 z = 1000
                 reordering = true
                 dragged = true
@@ -113,6 +116,7 @@ Item {
         }
 
         function doReordering() {
+            z = oldZ
             wrapper.reorder(_newIndex, _newFolderIndex)
             _newIndex = -1
         }
@@ -122,6 +126,7 @@ Item {
                 manager.startReorderTimer.stop()
             }
             if (reordering) {
+                z = oldZ
                 reordering = false
                 manager.reorderTimer.stop()
                 manager.stopScrolling()
diff --git a/usr/share/lipstick-jolla-home-qt5/notifications/NotificationItem.qml b/usr/share/lipstick-jolla-home-qt5/notifications/NotificationItem.qml
index 86be188..a8fd10a 100644
--- a/usr/share/lipstick-jolla-home-qt5/notifications/NotificationItem.qml
+++ b/usr/share/lipstick-jolla-home-qt5/notifications/NotificationItem.qml
@@ -12,7 +12,7 @@ import org.nemomobile.dbus 2.0
 import com.jolla.lipstick 0.1
 import org.nemomobile.lipstick 0.1
 
-Item {
+ListItem {
     id: root
 
     property alias showApplicationName: groupHeader.showName
@@ -61,6 +61,65 @@ Item {
     width: parent.width
     height: group.y + group.height
 
+    drag.target: !Lipstick.compositor.eventsLayer.housekeeping && hasRemovableMember ? root : null
+    drag.axis: Drag.XAxis
+    drag.minimumX: 0
+    drag.maximumX: width
+
+    drag.onActiveChanged: {
+        if (!drag.active) {
+            if (x > width / 3) {
+                state = "removing"
+            } else {
+                state = "default"
+            }
+        }
+    }
+
+    state: "default"
+
+    states: [
+        State {
+            name: "default"
+        },
+        State {
+            name: "dragging"
+            when: drag.active
+        },
+        State {
+            name: "removing"
+        }
+    ]
+
+    transitions: [
+        Transition {
+            from: "dragging"
+            to: "default"
+            NumberAnimation {
+                target: root
+                properties: "x"
+                to: 0
+                duration: 200
+            }
+        },
+        Transition {
+            from: "dragging"
+            to: "removing"
+            SequentialAnimation {
+                NumberAnimation {
+                    target: root
+                    properties: "x"
+                    to: width
+                    duration: 200
+                }
+                ScriptAction {
+                    script: root._notificationGroup.removeRequested()
+                }
+            }
+        }
+    ]
+
+
     NotificationGroupHeader {
         id: groupHeader
 
@@ -177,6 +236,64 @@ Item {
                     onTriggered: {
                         notification.actionInvoked("default")
                     }
+
+                    drag.target: !Lipstick.compositor.eventsLayer.housekeeping && userRemovable ? memberItem : null
+                    drag.axis: Drag.XAxis
+                    drag.minimumX: 0
+                    drag.maximumX: width
+
+                    drag.onActiveChanged: {
+                        if (!drag.active) {
+                            if (x > width / 3) {
+                                state = "removing"
+                            } else {
+                                state = "default"
+                            }
+                        }
+                    }
+
+                    state: "default"
+
+                    states: [
+                        State {
+                            name: "default"
+                        },
+                        State {
+                            name: "dragging"
+                            when: drag.active
+                        },
+                        State {
+                            name: "removing"
+                        }
+                    ]
+
+                    transitions: [
+                        Transition {
+                            from: "dragging"
+                            to: "default"
+                            NumberAnimation {
+                                target: memberItem
+                                properties: "x"
+                                to: 0
+                                duration: 200
+                            }
+                        },
+                        Transition {
+                            from: "dragging"
+                            to: "removing"
+                            SequentialAnimation {
+                                NumberAnimation {
+                                    target: memberItem
+                                    properties: "x"
+                                    to: width
+                                    duration: 200
+                                }
+                                ScriptAction {
+                                    script: notification.removeRequested()
+                                }
+                            }
+                        }
+                    ]
                 }
             }
         }
diff --git a/usr/share/lipstick-jolla-home-qt5/switcher/SwitcherItem.qml b/usr/share/lipstick-jolla-home-qt5/switcher/SwitcherItem.qml
index b6c281a..41f7498 100644
--- a/usr/share/lipstick-jolla-home-qt5/switcher/SwitcherItem.qml
+++ b/usr/share/lipstick-jolla-home-qt5/switcher/SwitcherItem.qml
@@ -28,7 +28,7 @@ EditableGridDelegate {
     property bool windowMapPending
     property int showingWid
     property int columns: 1
-    property bool closing: windowId != 0 && windowId == Lipstick.compositor.appLayer.closingWindowId
+    property bool closing: (windowId != 0 && windowId == Lipstick.compositor.appLayer.closingWindowId) || switcherItemRoot.drag.active
     property bool running
     property bool launching: pending
     property bool pending
@@ -191,6 +191,62 @@ EditableGridDelegate {
         }
     }
 
+    MouseArea {
+        id: switcherItemRoot
+        enabled: !wrapper.pending && !switcherRoot.closeApplicationEnabled && !closeAnimation.running
+        anchors.fill: parent
+        drag.target: wrapper.contentItem
+        drag.axis: Drag.XAxis
+        drag.minimumX: wrapper.x - width / 2
+        drag.maximumX: wrapper.x
+
+        property double dragDistance: wrapper.x - wrapper.contentItem.x
+        property double dragRatio: 2 * dragDistance / width
+        property int oldZ
+
+        drag.onActiveChanged: {
+            if (drag.active) {
+                oldZ = wrapper.contentItem.z
+                wrapper.contentItem.z = 1000
+            } else {
+                wrapper.contentItem.z = oldZ
+                if (dragDistance > width / 3) {
+                    wrapper.close()
+                } else {
+                    state = "default"
+                }
+            }
+        }
+
+        state: "default"
+
+        states: [
+            State {
+                name: "default"
+            },
+            State {
+                name: "dragging"
+                when: switcherItemRoot.drag.active
+            }
+        ]
+
+        transitions: Transition {
+            from: "dragging"
+            to: "default"
+            NumberAnimation {
+                target: wrapper.contentItem
+                properties: "x"
+                to: wrapper.x
+                duration: 150
+            }
+        }
+
+        Component.onCompleted: {
+            clicked.connect(wrapper.clicked)
+            pressAndHold.connect(wrapper.pressAndHold)
+        }
+    }
+
     LauncherIcon {
         size: Theme.iconSizeMedium
         anchors.centerIn: parent
@@ -374,6 +430,7 @@ EditableGridDelegate {
     Image {
         id: closingGraphic
         anchors.centerIn: parent
+        opacity: switcherItemRoot.drag.active ? switcherItemRoot.dragRatio : 1.0
         visible: wrapper.closing
         source: "image://theme/graphic-close-app"
         scale: wrapper.width / Theme.coverSizeLarge.width
